# **Equipo 1**

Yocelin Juárez Arroyo - A00819828

Luis Arturo Dan Fong - A01650672

Octavio Edelberto Reyes Torres - A00269987


#Tener todo en un csv para trabajar desde aqui

In [11]:
import warnings
warnings.filterwarnings("ignore")

from google.colab import drive
drive.mount('/content/drive')
#folder_path = '/content/drive/My Drive/Proyecto Integrador/'
#folder_path = '/content/drive/My Drive/Proyecto Integrador/colab/'

#Octavio - Luis
#folder_path = '/content/drive/My Drive/Proyecto Integrador/'

#Yocelin (owner)
folder_path = '/content/drive/My Drive/Proyecto Integrador/Proyecto Integrador/'

import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler, PowerTransformer, LabelEncoder
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
import pandas as pd
import gc

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [12]:
#Unir DF con una iniciativa Historia de éxito
ruta_archivo =  folder_path + 'bases/df_unido_limpioLGD_versionModelo.csv'
df_unidoLDG_encoded = pd.read_csv(ruta_archivo)

In [13]:
# Ajustar la configuración de pandas para mostrar todas las columnas
pd.set_option('display.max_columns', None)
# Mostrar las primeras filas del dataframe
display(df_unidoLDG_encoded.head())

categoria_dominio_correo_Educativo  categoria_dominio_correo_Genérico  \
0                               False                               True   
1                               False                               True   
2                               False                               True   
3                               False                               True   
4                               False                               True   

   categoria_dominio_correo_Otros  base_SEXO_Femenino  base_SEXO_Masculino  \
0                           False               False                 True   
1                           False               False                 True   
2                           False               False                 True   
3                           False               False                 True   
4                           False               False                 True   

   base_Tipo Escuela_Arquitectura y Diseño  \
0                                    False   
1                                    False   
2                                    False   
3                                    False   
4                                    False   

   base_Tipo Escuela_Gobierno y Ciencias Sociales  \
0                                           False   
1                                           False   
2                                           False   
3                                           False   
4                                           False   

   base_Tipo Escuela_Humanidades y Educación  \
0                                      False   
1                                      False   
2                                      False   
3                                      False   
4                                      False   

   base_Tipo Escuela_Ingeniería y Ciencias  \
0                                    False   
1                                    False   
2                                    False   
3                                    False   
4                                    False   

   base_Tipo Escuela_Medicina y Ciencias de la salud  \
0                                              False   
1                                              False   
2                                              False   
3                                              False   
4                                              False   

   base_Tipo Escuela_Negocios  base_IDCAMPUS_   base_IDCAMPUS_1  \
0                        True            False            False   
1                        True            False            False   
2                        True            False            False   
3                        True            False            False   
4                        True            False            False   

   base_IDCAMPUS_2  base_IDCAMPUS_8  base_IDCAMPUS_A  base_IDCAMPUS_A1  \
0            False            False             True             False   
1            False            False             True             False   
2            False            False            False             False   
3            False            False            False             False   
4            False            False            False             False   

   base_IDCAMPUS_AA  base_IDCAMPUS_C  base_IDCAMPUS_D  base_IDCAMPUS_E  \
0             False            False            False            False   
1             False            False            False            False   
2             False            False            False            False   
3             False            False            False            False   
4             False            False            False            False   

   base_IDCAMPUS_F  base_IDCAMPUS_G  base_IDCAMPUS_G1  base_IDCAMPUS_GA  \
0            False            False             False             False   
1            False            False             False             False   
2            False            False             False       

#**Avance 4. Modelos alternativos**

##6 Modelos

Se construyeron 6 modelos diferentes que ejecutamos para cada una de las variables objetivo definidas en nuestro proyecto ('es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito').

De los cuales evaluaremos su rendimiento mediante métricas clave, como **accuracy**, **F1-score** y **recall**.
Siendo Accuracy la métrica principal y considerando los **tiempos de entrenamiento**.

Los modelos a utilizar son:

**Random Forest:** Modelo basado en múltiples árboles de decisión que mejora la precisión al promediar predicciones, ideal para capturar relaciones complejas en los datos.

**K-Nearest Neighbors** (KNN): Clasificador que predice la clase de una observación en función de sus vecinos más cercanos en el espacio de características, útil en problemas donde la cercanía de los datos es relevante.

**Ridge:** Modelo de regresión lineal con regularización que ayuda a controlar el sobreajuste al penalizar coeficientes altos, ideal para manejar datos con colinealidad.

**Lasso:** Similar a Ridge, este modelo de regresión lineal agrega regularización para reducir coeficientes a cero en algunos casos, ayudando en la selección de características relevantes.

**Decision Tree:** Algoritmo de clasificación que utiliza una estructura de árbol para dividir los datos en función de características, fácil de interpretar y capaz de captar relaciones complejas.

**Regresión Logística:** Modelo estadístico de clasificación que estima la probabilidad de que una observación pertenezca a una clase específica, ideal para problemas de clasificación binaria.


###Random Forest 3 variables objetivo

In [17]:
# Importar librerías necesarias
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, f1_score, recall_score
import time
from IPython.display import display

# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_unidoLDG_encoded.drop(columns=objective_variables)
    y = df_unidoLDG_encoded[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo de Random Forest con una profundidad máxima de 3
    model = RandomForestClassifier(n_estimators=100, max_depth=3, random_state=42)
    model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = model.predict(X_test)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')

    # Mostrar los resultados
    display(f"Tiempo de entrenamiento del modelo: {training_time:.2f} segundos")
    display(f"Accuracy del modelo Random Forest con max_depth=3 para '{objective}': {accuracy:.4f}")
    display(f"F1-Score del modelo: {f1:.4f}")
    display(f"Recall del modelo: {recall:.4f}")
    display("\nReporte de clasificación completo:")
    display(classification_report(y_test, y_pred))

"\nResultados para el modelo con variable objetivo 'es_LDGincubadora':"

'Tiempo de entrenamiento del modelo: 10.92 segundos'

"Accuracy del modelo Random Forest con max_depth=3 para 'es_LDGincubadora': 0.9003"

'F1-Score del modelo: 0.8990'

'Recall del modelo: 0.9003'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.87      0.96      0.92     38723\n           1       0.94      0.81      0.87     28752\n\n    accuracy                           0.90     67475\n   macro avg       0.91      0.89      0.90     67475\nweighted avg       0.90      0.90      0.90     67475\n'

"\nResultados para el modelo con variable objetivo 'es_Voluntariado':"

'Tiempo de entrenamiento del modelo: 10.92 segundos'

"Accuracy del modelo Random Forest con max_depth=3 para 'es_Voluntariado': 0.9003"

'F1-Score del modelo: 0.8980'

'Recall del modelo: 0.9003'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.86      0.99      0.92     38609\n           1       0.99      0.77      0.87     28866\n\n    accuracy                           0.90     67475\n   macro avg       0.92      0.88      0.89     67475\nweighted avg       0.91      0.90      0.90     67475\n'

"\nResultados para el modelo con variable objetivo 'es_HistoriaExito':"

'Tiempo de entrenamiento del modelo: 11.66 segundos'

"Accuracy del modelo Random Forest con max_depth=3 para 'es_HistoriaExito': 0.9024"

'F1-Score del modelo: 0.9000'

'Recall del modelo: 0.9024'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.85      1.00      0.92     38381\n           1       1.00      0.77      0.87     29094\n\n    accuracy                           0.90     67475\n   macro avg       0.93      0.89      0.90     67475\nweighted avg       0.92      0.90      0.90     67475\n'

###KNN 3 variables objetivo

In [16]:
# Importar las librerías necesarias
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, f1_score, recall_score
import time
from IPython.display import display

# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_unidoLDG_encoded.drop(columns=objective_variables)
    y = df_unidoLDG_encoded[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo KNN
    knn_model = KNeighborsClassifier(n_neighbors=3, algorithm='ball_tree')
    knn_model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = knn_model.predict(X_test)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')

    # Mostrar los resultados
    display(f"Tiempo de entrenamiento del modelo: {training_time:.2f} segundos")
    display(f"Accuracy del modelo KNN para '{objective}': {accuracy:.4f}")
    display(f"F1-Score del modelo: {f1:.4f}")
    display(f"Recall del modelo: {recall:.4f}")
    display("\nReporte de clasificación completo:")
    display(classification_report(y_test, y_pred))

"\nResultados para el modelo con variable objetivo 'es_LDGincubadora':"

'Tiempo de entrenamiento del modelo: 19.42 segundos'

"Accuracy del modelo KNN para 'es_LDGincubadora': 0.9999"

'F1-Score del modelo: 0.9999'

'Recall del modelo: 0.9999'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       1.00      1.00      1.00     38723\n           1       1.00      1.00      1.00     28752\n\n    accuracy                           1.00     67475\n   macro avg       1.00      1.00      1.00     67475\nweighted avg       1.00      1.00      1.00     67475\n'

"\nResultados para el modelo con variable objetivo 'es_Voluntariado':"

'Tiempo de entrenamiento del modelo: 17.88 segundos'

"Accuracy del modelo KNN para 'es_Voluntariado': 0.9999"

'F1-Score del modelo: 0.9999'

'Recall del modelo: 0.9999'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       1.00      1.00      1.00     38609\n           1       1.00      1.00      1.00     28866\n\n    accuracy                           1.00     67475\n   macro avg       1.00      1.00      1.00     67475\nweighted avg       1.00      1.00      1.00     67475\n'

"\nResultados para el modelo con variable objetivo 'es_HistoriaExito':"

'Tiempo de entrenamiento del modelo: 19.00 segundos'

"Accuracy del modelo KNN para 'es_HistoriaExito': 1.0000"

'F1-Score del modelo: 1.0000'

'Recall del modelo: 1.0000'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       1.00      1.00      1.00     38381\n           1       1.00      1.00      1.00     29094\n\n    accuracy                           1.00     67475\n   macro avg       1.00      1.00      1.00     67475\nweighted avg       1.00      1.00      1.00     67475\n'

###Ridge 3 variables objetivo

In [ ]:
# Importar las librerías necesarias
from sklearn.linear_model import Ridge
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score, recall_score, classification_report
import time
import numpy as np
from IPython.display import display

# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_unidoLDG_encoded.drop(columns=objective_variables)
    y = df_unidoLDG_encoded[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo de Ridge
    ridge_model = Ridge(alpha=1.0)  # Ajusta alpha según sea necesario
    ridge_model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = ridge_model.predict(X_test)

    # Convertir las predicciones continuas en clases binarias (0 o 1)
    y_pred_class = np.where(y_pred >= 0.5, 1, 0)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred_class)
    f1 = f1_score(y_test, y_pred_class, average='weighted')
    recall = recall_score(y_test, y_pred_class, average='weighted')

    # Mostrar los resultados
    display(f"Tiempo de entrenamiento del modelo: {training_time:.2f} segundos")
    display(f"Accuracy del modelo Ridge para '{objective}': {accuracy:.4f}")
    display(f"F1-Score del modelo: {f1:.4f}")
    display(f"Recall del modelo: {recall:.4f}")
    display("\nReporte de clasificación completo:")
    display(classification_report(y_test, y_pred_class))


"\nResultados para el modelo con variable objetivo 'es_LDGincubadora':"

'Tiempo de entrenamiento del modelo: 3.61 segundos'

"Accuracy del modelo Ridge para 'es_LDGincubadora': 0.9740"

'F1-Score del modelo: 0.9740'

'Recall del modelo: 0.9740'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.98      0.98      0.98     38723\n           1       0.97      0.97      0.97     28752\n\n    accuracy                           0.97     67475\n   macro avg       0.97      0.97      0.97     67475\nweighted avg       0.97      0.97      0.97     67475\n'

"\nResultados para el modelo con variable objetivo 'es_Voluntariado':"

'Tiempo de entrenamiento del modelo: 2.45 segundos'

"Accuracy del modelo Ridge para 'es_Voluntariado': 0.9750"

'F1-Score del modelo: 0.9750'

'Recall del modelo: 0.9750'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.99      0.97      0.98     38609\n           1       0.96      0.99      0.97     28866\n\n    accuracy                           0.97     67475\n   macro avg       0.97      0.98      0.97     67475\nweighted avg       0.98      0.97      0.97     67475\n'

"\nResultados para el modelo con variable objetivo 'es_HistoriaExito':"

'Tiempo de entrenamiento del modelo: 2.01 segundos'

"Accuracy del modelo Ridge para 'es_HistoriaExito': 0.9896"

'F1-Score del modelo: 0.9896'

'Recall del modelo: 0.9896'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.99      0.99      0.99     38381\n           1       0.98      0.99      0.99     29094\n\n    accuracy                           0.99     67475\n   macro avg       0.99      0.99      0.99     67475\nweighted avg       0.99      0.99      0.99     67475\n'

###Lasso 3 variables objetivo

In [ ]:
# Importar las librerías necesarias
from sklearn.linear_model import Lasso
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score, recall_score, classification_report
import time
import numpy as np
from IPython.display import display

# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_unidoLDG_encoded.drop(columns=objective_variables)
    y = df_unidoLDG_encoded[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo de Lasso
    lasso_model = Lasso(alpha=1.0)  # Puedes ajustar el valor de alpha para regularización
    lasso_model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = lasso_model.predict(X_test)

    # Convertir las predicciones continuas en clases binarias (0 o 1)
    y_pred_class = np.where(y_pred >= 0.5, 1, 0)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred_class)
    f1 = f1_score(y_test, y_pred_class, average='weighted')
    recall = recall_score(y_test, y_pred_class, average='weighted')

    # Mostrar los resultados
    display(f"Tiempo de entrenamiento del modelo: {training_time:.2f} segundos")
    display(f"Accuracy del modelo Lasso para '{objective}': {accuracy:.4f}")
    display(f"F1-Score del modelo: {f1:.4f}")
    display(f"Recall del modelo: {recall:.4f}")
    display("\nReporte de clasificación completo:")
    display(classification_report(y_test, y_pred_class))


"\nResultados para el modelo con variable objetivo 'es_LDGincubadora':"

'Tiempo de entrenamiento del modelo: 5.38 segundos'

"Accuracy del modelo Lasso para 'es_LDGincubadora': 0.8295"

'F1-Score del modelo: 0.8273'

'Recall del modelo: 0.8295'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.82      0.90      0.86     38723\n           1       0.85      0.73      0.79     28752\n\n    accuracy                           0.83     67475\n   macro avg       0.83      0.82      0.82     67475\nweighted avg       0.83      0.83      0.83     67475\n'

"\nResultados para el modelo con variable objetivo 'es_Voluntariado':"

'Tiempo de entrenamiento del modelo: 4.76 segundos'

"Accuracy del modelo Lasso para 'es_Voluntariado': 0.7195"

'F1-Score del modelo: 0.6947'

'Recall del modelo: 0.7195'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.69      0.94      0.79     38609\n           1       0.84      0.42      0.56     28866\n\n    accuracy                           0.72     67475\n   macro avg       0.77      0.68      0.68     67475\nweighted avg       0.75      0.72      0.69     67475\n'

"\nResultados para el modelo con variable objetivo 'es_HistoriaExito':"

'Tiempo de entrenamiento del modelo: 3.72 segundos'

"Accuracy del modelo Lasso para 'es_HistoriaExito': 0.6433"

'F1-Score del modelo: 0.6225'

'Recall del modelo: 0.6433'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.64      0.84      0.73     38381\n           1       0.64      0.39      0.48     29094\n\n    accuracy                           0.64     67475\n   macro avg       0.64      0.61      0.61     67475\nweighted avg       0.64      0.64      0.62     67475\n'

###Decision Tree 3 variables objetivo

In [ ]:
# Importar las librerías necesarias
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, f1_score, recall_score, classification_report
import time
from IPython.display import display

# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_unidoLDG_encoded.drop(columns=objective_variables)
    y = df_unidoLDG_encoded[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo de Árbol de Decisión
    model = DecisionTreeClassifier(max_depth=5, min_samples_split=10, min_samples_leaf=5)
    model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = model.predict(X_test)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')

    # Mostrar los resultados
    display(f"Tiempo de entrenamiento del modelo: {training_time:.2f} segundos")
    display(f"Accuracy del modelo de Árbol de Decisión para '{objective}': {accuracy:.4f}")
    display(f"F1-Score del modelo: {f1:.4f}")
    display(f"Recall del modelo: {recall:.4f}")
    display("\nReporte de clasificación completo:")
    display(classification_report(y_test, y_pred))


###Regresión Logística 3 variables objetivo

In [ ]:
# Importar las librerías necesarias
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score, recall_score, classification_report
import time
from IPython.display import display

# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# Tomar una muestra reducida del DataFrame para evitar carga excesiva
df_sample = df_unidoLDG_encoded.sample(frac=0.7, random_state=42)  # Reduce la muestra al 30% de los datos

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_sample.drop(columns=objective_variables)
    y = df_sample[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo de Regresión Logística
    model = LogisticRegression(max_iter=500)
    model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = model.predict(X_test)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')

    # Mostrar los resultados
    display(f"Tiempo de entrenamiento del modelo: {training_time:.2f} segundos")
    display(f"Accuracy del modelo de Regresión Logística para '{objective}': {accuracy:.4f}")
    display(f"F1-Score del modelo: {f1:.4f}")
    display(f"Recall del modelo: {recall:.4f}")
    display("\nReporte de clasificación completo:")
    display(classification_report(y_test, y_pred))


"\nResultados para el modelo con variable objetivo 'es_LDGincubadora':"

'Tiempo de entrenamiento del modelo: 69.57 segundos'

"Accuracy del modelo de Regresión Logística para 'es_LDGincubadora': 0.9468"

'F1-Score del modelo: 0.9468'

'Recall del modelo: 0.9468'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.95      0.96      0.95     27005\n           1       0.94      0.93      0.94     20227\n\n    accuracy                           0.95     47232\n   macro avg       0.95      0.94      0.95     47232\nweighted avg       0.95      0.95      0.95     47232\n'

"\nResultados para el modelo con variable objetivo 'es_Voluntariado':"

'Tiempo de entrenamiento del modelo: 66.77 segundos'

"Accuracy del modelo de Regresión Logística para 'es_Voluntariado': 0.9387"

'F1-Score del modelo: 0.9387'

'Recall del modelo: 0.9387'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.95      0.95      0.95     26911\n           1       0.93      0.93      0.93     20321\n\n    accuracy                           0.94     47232\n   macro avg       0.94      0.94      0.94     47232\nweighted avg       0.94      0.94      0.94     47232\n'

"\nResultados para el modelo con variable objetivo 'es_HistoriaExito':"

'Tiempo de entrenamiento del modelo: 71.66 segundos'

"Accuracy del modelo de Regresión Logística para 'es_HistoriaExito': 0.9825"

'F1-Score del modelo: 0.9825'

'Recall del modelo: 0.9825'

'\nReporte de clasificación completo:'

'              precision    recall  f1-score   support\n\n           0       0.99      0.98      0.98     26776\n           1       0.97      0.99      0.98     20456\n\n    accuracy                           0.98     47232\n   macro avg       0.98      0.98      0.98     47232\nweighted avg       0.98      0.98      0.98     47232\n'

##Comparativa de los 6 modelos

Extraímos los valores de accuracy, f1-score, recall y el tiempo de entrenamiento para cada modelo evaluado. Esto nos permitió analizar el rendimiento de cada uno en relación con las métricas principales.

 ordenamos la tabla según la métrica principal (accuracy) para identificar fácilmente los modelos más efectivos.

In [17]:
# Resultados para cada modelo
model_names = ["Random Forest", "KNN", "Ridge", "Lasso", "Decision Tree", "Logistic Regression"]
accuracies = [0.92, 1.00, 0.99, 0.64 ,  0.99, 0.98]
f1_scores = [0.90, 1.00, 0.99, 0.64,  0.99, 0.98]
recalls = [0.90, 1.00, 0.99, 0.64,  0.99, 0.98]
training_times = [13.77, 21.52, 2.24, 5.77, 3.29, 67.23]  # en segundos

# Crear un DataFrame con los resultados
results_df = pd.DataFrame({
    "Modelo": model_names,
    "Accuracy": accuracies,
    "F1-Score": f1_scores,
    "Recall": recalls,
    "Tiempo de Entrenamiento (s)": training_times
})

# Ordenar el DataFrame por la métrica principal (en este caso, 'Accuracy')
results_df = results_df.sort_values(by="Accuracy", ascending=False)

# Mostrar la tabla consolidada
display("Tabla Comparativa de Modelos:")
display(results_df)


'Tabla Comparativa de Modelos:'

Modelo  Accuracy  F1-Score  Recall  \
1                  KNN      1.00      1.00    1.00   
2                Ridge      0.99      0.99    0.99   
4        Decision Tree      0.99      0.99    0.99   
5  Logistic Regression      0.98      0.98    0.98   
0        Random Forest      0.92      0.90    0.90   
3                Lasso      0.64      0.64    0.64   

   Tiempo de Entrenamiento (s)  
1                        21.52  
2                         2.24  
4                         3.29  
5                        67.23  
0                        13.77  
3                         5.77

####Análisis de Resultados:

Al analizar la tabla comparativa, observamos que el modelo KNN alcanzó un rendimiento perfecto en todas las métricas de accuracy, f1-score, y recall, con valores de 1.00 en cada una. Este modelo destaca claramente como el más preciso, aunque su tiempo de entrenamiento de 21.52 segundos es considerable en comparación con otros modelos que también muestran un rendimiento sobresaliente.

El modelo Ridge y el Decision Tree también obtuvieron resultados muy altos, con valores de 0.99 en todas las métricas y se entrenaron en tiempos mucho menores que el KNN (2.24 y 3.29 segundos, respectivamente).

Logistic Regression obtuvo valores levemente inferiores, con un accuracy, f1-score, y recall de 0.98. Aunque este modelo es bastante preciso, su tiempo de entrenamiento es significativamente mayor (67.23 segundos).

El Random Forest, con un accuracy de 0.92, se desempeñó bien en general, aunque fue superado en precisión y tiempo por otros modelos como Ridge y Decision Tree.

Por último, el modelo Lasso mostró el rendimiento más bajo, con métricas de 0.64, lo que indica que no es la mejor opción para este problema específico.

En conclusión, los modelos KNN, Ridge, y Decision Tree destacaron por su alta precisión, siendo **Ridge y Decision Tree** los más eficientes en términos de tiempo de entrenamiento. Estos resultados nos ayudarán a elegir los mejores modelos para el ajuste fino y la optimización final.

###Modelos con Mejor Desempeño

Para asegurar que nuestro modelo no esté sobreajustado, seleccionamos el modelo Ridge y el árbol de decisión con una profundidad controlada.

**Modelo Ridge:**
* **Ventaja:** Ridge es una técnica de regresión lineal regularizada que penaliza los coeficientes de los predictores, lo cual nos ayuda a reducir el riesgo de sobreajuste. Gracias a su regularización, controlada por el hiperparámetro alpha, evitamos que los coeficientes crezcan excesivamente, lo que permite manejar mejor los datos ruidosos y minimizar el sobreajuste.
* **Rendimiento y Eficiencia:** Este modelo obtuvo un accuracy de 0.99 en las métricas clave, con un tiempo de entrenamiento bajo (2.24 segundos). Este equilibrio entre precisión y robustez ante el sobreajuste hace que el modelo Ridge sea una opción sólida.

**Árbol de Decisión** (con una profundidad limitada):
* **Ventaja:** Aunque los árboles de decisión tienen tendencia a sobreajustarse, establecer una profundidad máxima (en este caso, configurando max_depth=5) nos ayuda a evitarlo. Con una profundidad controlada, el árbol puede captar patrones generales sin volverse excesivamente específico con los datos de entrenamiento.
* **Rendimiento:** Este modelo obtuvo métricas de 0.99 en accuracy, f1-score y recall, con un tiempo de entrenamiento razonable (3.29 segundos). Un árbol de decisión bien regulado es, por tanto, una opción adecuada para evitar el sobreajuste, además de que mantiene la interpretabilidad.

##Ajuste Fino en Ridge

Optimizamos el modelo Ridge mediante un proceso de ajuste fino utilizando **RandomizedSearchCV**, una herramienta que nos permite probar de manera eficiente combinaciones aleatorias de hiperparámetros para maximizar el rendimiento del modelo, en este caso medido por el Mean Squared Error (MSE) negativo.

**Muestra y División de Datos:**
Primero, trabajamos con una muestra del 30% del conjunto de datos original para reducir el tiempo de procesamiento. Los datos se dividen en conjuntos de entrenamiento y prueba, y se limpian valores NaN o infinitos para asegurar la estabilidad del modelo.

**Rango de Hiperparámetros** (param_distributions):
Definimos un rango reducido de valores para los hiperparámetros clave de Ridge:

* alpha: Controla el nivel de regularización y se ajusta en una escala logarítmica entre 10^-2 y 10^1, lo que nos ayuda a cubrir un rango representativo de regularización.
* solver: Se establece en auto para permitir que el modelo seleccione el mejor método de optimización según los datos.

Con 5 combinaciones aleatorias de hiperparámetros y validación cruzada de 3 pliegues, identificamos los valores óptimos para maximizar el rendimiento en el conjunto de prueba.

In [6]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import RandomizedSearchCV, train_test_split
import numpy as np

# Crear una muestra del DataFrame
df_sample = df_unidoLDG_encoded.sample(frac=0.3, random_state=42)

# Definir las variables predictoras (X) y la variable objetivo (y) en la muestra
X_sample = df_sample.drop(columns=['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito'])
y_sample = df_sample['es_LDGincubadora']

# Dividir los datos de la muestra en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X_sample, y_sample, test_size=0.2, random_state=42)

# Verificar y limpiar NaN o infinitos en los datos
X_train = X_train.replace([np.inf, -np.inf], np.nan).fillna(0)
y_train = y_train.replace([np.inf, -np.inf], np.nan).fillna(y_train.mean())

# Definir el modelo Ridge
ridge_model = Ridge()

# Definir un rango reducido de valores para los hiperparámetros
param_distributions = {
    'alpha': np.logspace(-2, 1, 3),  # Rango más limitado para alpha
    'solver': ['auto']               # Un solo solver
}

# Configurar la búsqueda aleatoria (Randomized Search) con menos iteraciones
random_search = RandomizedSearchCV(
    estimator=ridge_model,
    param_distributions=param_distributions,
    n_iter=5,             # Menor número de combinaciones aleatorias a probar
    scoring='neg_mean_squared_error',  # Métrica para regresión
    cv=3,
    n_jobs=-1,
    random_state=42
)

# Entrenar el Randomized Search en los datos de entrenamiento
random_search.fit(X_train, y_train)

# Mostrar los mejores hiperparámetros y el rendimiento correspondiente
display("Mejores hiperparámetros encontrados:", random_search.best_params_)
display("Mejor rendimiento (neg_mean_squared_error) obtenido:", random_search.best_score_)

'Mejores hiperparámetros encontrados:'

{'solver': 'auto', 'alpha': 0.31622776601683794}

'Mejor rendimiento (neg_mean_squared_error) obtenido:'

-0.030760871557608534

Entrenaremos el modelo final de Ridge utilizando los mejores hiperparámetros encontrados en el ajuste fino (alpha=0.316 y solver='auto').

Primero, ajustaremos el modelo a los datos de entrenamiento (X_train, y_train) para que aprenda de ellos.

A continuación, realizaremos predicciones en el conjunto de prueba (X_test) y evaluaremos el rendimiento del modelo en estos datos no vistos. Para medir el rendimiento, usaremos métricas de regresión como:

* Mean Squared Error (MSE): Evalúa el error promedio al cuadrado entre las predicciones y los valores reales.
* Mean Absolute Error (MAE): Mide el error promedio absoluto.
* R² Score: Indica qué tan bien el modelo explica la variabilidad de los datos.

In [7]:
# Entrenar el modelo final con los mejores hiperparámetros
final_ridge_model = Ridge(alpha=0.316, solver='auto')
final_ridge_model.fit(X_train, y_train)

# Predecir en el conjunto de prueba
y_pred = final_ridge_model.predict(X_test)

# Evaluar el rendimiento en el conjunto de prueba
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

display("Rendimiento en el conjunto de prueba:")
display("Mean Squared Error (MSE):", mse)
display("Mean Absolute Error (MAE):", mae)
display("R^2 Score:", r2)

'Rendimiento en el conjunto de prueba:'

'Mean Squared Error (MSE):'

0.0300062082810421

'Mean Absolute Error (MAE):'

0.09660605685854118

'R^2 Score:'

0.8773762944738238

Resultados del Rendimiento en el Conjunto de Prueba

**Mean Squared Error (MSE): 0.0300**

* El MSE representa el error promedio al cuadrado entre las predicciones y los valores reales, y un valor bajo indica que las predicciones están, en general, cerca de los valores reales.
En este caso, el MSE de 0.0300 sugiere que el modelo tiene un error promedio relativamente pequeño, lo cual es un buen indicador de precisión.

**Mean Absolute Error (MAE): 0.0966**

* El MAE muestra el error promedio absoluto, es decir, el promedio de las diferencias absolutas entre las predicciones y los valores reales.
Un MAE de 0.0966 significa que, en promedio, las predicciones del modelo están a aproximadamente 0.0966 unidades de los valores reales. Este valor también indica un rendimiento sólido, con desviaciones relativamente pequeñas.

**R² Score: 0.8774**

* El R² Score, o coeficiente de determinación, mide la proporción de la varianza en la variable objetivo que el modelo puede explicar. Un valor de 1 significa que el modelo explica perfectamente los datos, mientras que un valor cercano a 0 indica que no explica la variabilidad.
En este caso, el R² de 0.8774 indica que el modelo explica aproximadamente el 87.7% de la variabilidad en los datos de prueba, lo cual es un buen nivel de ajuste, mostrando que el modelo generaliza bien en datos nuevos.

##Ajuste Fino en Árbol de Decisión

###Hiperparámetros a Ajustar para el Árbol de Decisión

Los hiperparámetros más importantes para un árbol de decisión incluyen:

* max_depth: La profundidad máxima del árbol. Limitar este valor ayuda a evitar el sobreajuste.
* min_samples_split: El número mínimo de muestras requeridas para dividir un nodo. Valores más altos evitan que el árbol crezca demasiado en profundidad.
* min_samples_leaf: El número mínimo de muestras que debe tener un nodo hoja. Ayuda a reducir el sobreajuste.
* max_features: El número máximo de características a considerar al dividir un nodo. Puede ser un valor entero, un porcentaje ("sqrt" o "log2"), o None para usar todas las características.

Vamos a utilizar RandomizedSearchCV para explorar el espacio de hiperparámetros, lo que permitirá reducir el tiempo de búsqueda en comparación con GridSearchCV.

In [8]:
from sklearn.tree import DecisionTreeRegressor  # Usa DecisionTreeClassifier si es clasificación
from sklearn.model_selection import RandomizedSearchCV, train_test_split
import numpy as np

# Definir las variables predictoras (X) y la variable objetivo (y)
X = df_unidoLDG_encoded.drop(columns=['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito'])
y = df_unidoLDG_encoded['es_LDGincubadora']  # Variable objetivo seleccionada

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Definir el modelo de Árbol de Decisión
decision_tree_model = DecisionTreeRegressor(random_state=42)

# Definir el rango de valores para los hiperparámetros
param_distributions = {
    'max_depth': [None, 10, 20, 30, 40, 50],      # Profundidad máxima del árbol
    'min_samples_split': [2, 5, 10, 20],          # Número mínimo de muestras para dividir un nodo
    'min_samples_leaf': [1, 2, 5, 10],            # Número mínimo de muestras en una hoja
    'max_features': ['auto', 'sqrt', 'log2', None] # Máximo de características a considerar
}

# Configurar la búsqueda aleatoria (Randomized Search) con un número limitado de iteraciones
random_search = RandomizedSearchCV(
    estimator=decision_tree_model,
    param_distributions=param_distributions,
    n_iter=10,             # Número de combinaciones aleatorias a probar
    scoring='neg_mean_squared_error',  # Para regresión; usa 'accuracy' si es clasificación
    cv=3,                  # Validación cruzada de 3 pliegues para reducir tiempo
    n_jobs=-1,             # Usa todos los núcleos disponibles para acelerar el proceso
    random_state=42
)

# Entrenar el Randomized Search en los datos de entrenamiento
random_search.fit(X_train, y_train)

# Mostrar los mejores hiperparámetros y el rendimiento correspondiente
display("Mejores hiperparámetros encontrados:", random_search.best_params_)
display("Mejor rendimiento (neg_mean_squared_error) obtenido:", random_search.best_score_)


'Mejores hiperparámetros encontrados:'

{'min_samples_split': 2,
 'min_samples_leaf': 2,
 'max_features': None,
 'max_depth': 50}

'Mejor rendimiento (neg_mean_squared_error) obtenido:'

-0.00035061019204361095

####Interpretación de los Resultados



**Mejor Rendimiento:**

neg_mean_squared_error: -0.00035
Dado que RandomizedSearchCV devuelve el valor negativo del MSE, esto significa que el Mean Squared Error (MSE) del modelo en el conjunto de validación es aproximadamente 0.00035. Este es un valor bastante bajo, lo que indica que el modelo tiene un buen ajuste en los datos de entrenamiento y es capaz de predecir con precisión.

**Los hiperparámetros optimizados indican que:**

* Profundidad del Árbol (max_depth=50): Permitir que el árbol crezca hasta una profundidad de 50 indica que el modelo necesita una estructura más detallada para capturar las relaciones en los datos.
* Control de División y Tamaño de las Hojas (min_samples_split=2 y min_samples_leaf=2): Estas configuraciones permiten que el modelo capture patrones con precisión, evitando la creación de nodos con muy pocas muestras y reduciendo el riesgo de sobreajuste.

####Entrenar y Evaluar el Modelo Final

In [9]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# Entrenar el modelo final con los mejores hiperparámetros
final_decision_tree_model = DecisionTreeRegressor(
    min_samples_split=2,
    min_samples_leaf=2,
    max_features=None,
    max_depth=50,
    random_state=42
)

final_decision_tree_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = final_decision_tree_model.predict(X_test)

# Calcular las métricas de evaluación
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

display("Rendimiento en el conjunto de prueba:")
display("Mean Squared Error (MSE):", mse)
display("Mean Absolute Error (MAE):", mae)
display("R^2 Score:", r2)


'Rendimiento en el conjunto de prueba:'

'Mean Squared Error (MSE):'

0.00019254023431610308

'Mean Absolute Error (MAE):'

0.00023749935260769623

'R^2 Score:'

0.9992126456585579

Resultados del Rendimiento en el Conjunto de Prueba

**Mean Squared Error (MSE): 0.00019**

* El MSE es extremadamente bajo, lo que indica que el modelo tiene un error promedio al cuadrado muy pequeño en sus predicciones. Esto sugiere que las predicciones están, en general, muy cercanas a los valores reales.

**Mean Absolute Error (MAE): 0.00024**

* El MAE, que representa el error promedio absoluto, también es muy bajo. Un MAE de 0.00024 implica que, en promedio, las predicciones del modelo difieren de los valores reales por solo 0.00024 unidades, lo cual es excelente y muestra una gran precisión.

**R² Score: 0.9992**

* El R² Score de 0.9992 significa que el modelo explica el 99.92% de la variabilidad de los datos en el conjunto de prueba. Este es un valor muy alto, lo cual indica que el modelo generaliza muy bien y que es capaz de capturar casi toda la variabilidad de los datos.

##Comparación de Resultados

In [10]:
import pandas as pd

# Crear el DataFrame con los datos de la tabla
data = {
    "Métrica": ["Mean Squared Error (MSE)", "Mean Absolute Error (MAE)", "R² Score", "Tiempo de Entrenamiento"],
    "Árbol de Decisión": [0.00019, 0.00024, 0.9992, "Generalmente mayor que Ridge"],
    "Ridge": [0.0300, 0.0966, 0.8774, "Generalmente más rápido"]
}

# Crear el DataFrame
comparison_df = pd.DataFrame(data)

# Mostrar la tabla
display(comparison_df)


Métrica             Árbol de Decisión  \
0   Mean Squared Error (MSE)                       0.00019   
1  Mean Absolute Error (MAE)                       0.00024   
2                   R² Score                        0.9992   
3    Tiempo de Entrenamiento  Generalmente mayor que Ridge   

                     Ridge  
0                     0.03  
1                   0.0966  
2                   0.8774  
3  Generalmente más rápido

####Análisis de los Resultados

**Precisión:**

El Árbol de Decisión supera ampliamente al modelo Ridge en todas las métricas de rendimiento en el conjunto de prueba:

* MSE y MAE son significativamente más bajos en el Árbol de Decisión, lo cual indica un menor error promedio en sus predicciones.
* El R² Score de 0.9992 en el Árbol de Decisión sugiere que este modelo explica prácticamente toda la variabilidad en los datos de prueba, mientras que Ridge alcanza un R² de 0.8774.

**Tiempo de Entrenamiento:**

* El modelo Ridge es generalmente más rápido en el entrenamiento debido a su naturaleza lineal. El Árbol de Decisión, especialmente con una profundidad máxima alta como la seleccionada (max_depth=50), puede tomar más tiempo en entrenarse.
* Sin embargo, dado el excelente rendimiento del Árbol de Decisión en precisión, este tiempo de entrenamiento adicional parece justificado.

**Interpretabilidad:**

* El modelo Ridge tiene una ventaja en interpretabilidad, ya que los modelos lineales permiten entender de manera más directa la relación entre las variables predictoras y el objetivo.
* El Árbol de Decisión es menos interpretable a altas profundidades, pero sigue siendo más interpretable que otros modelos complejos como los bosques aleatorios o las redes neuronales.

**Estabilidad:**

* Los árboles de decisión pueden ser menos estables (pequeñas variaciones en los datos pueden cambiar la estructura del árbol).
* Ridge, por su naturaleza lineal y regularizada, suele ofrecer predicciones más estables y puede ser preferible en problemas donde la estabilidad es fundamental.

**Selección Final:**
Dado el rendimiento superior del Árbol de Decisión en precisión y su capacidad para captar patrones complejos en los datos, seleccionamos el **Árbol de Decisión como el modelo final**.